### PR TITLE
Add tests true continuum

### PR DESCRIPTION
**Pull request by cramirezpe**
Added true continuum tests (test data not yet shared as it contains desi mock data).

### CODE REVIEW

File: py/picca/tests/delta_extraction/expected_flux_tests.py [hunk 1]
```diff
@@ -805,6 +812,442 @@ def test_expected_flux(self):
...
+        with self.assertRaises(ExpectedFluxError):
+            setup_forest("lin", pixel_step=0.4)
+            expected_flux = TrueContinuum(config["expected flux"])
+
+    def test_true_continuum_read_raw_statistics(self):
+        """Test reading raw statistics files"""
+        # setup Forest variables; case: linear wavelength solution
+        setup_forest("lin", pixel_step=2.4)
+
+        data_dir = Path(__file__).parent / "data"
```
Comment (iprafols) [inline py/picca/tests/delta_extraction/expected_flux_tests.py:871]: use variable `THIS_DIR` to access the directories. Not sure where the path is, but you could define another variable similar to `THIS_DIR` at the top that points to the right place

File: py/picca/tests/delta_extraction/expected_flux_tests.py [hunk 2]
```diff
@@ -805,6 +812,442 @@ def test_expected_flux(self):
...
+
+        for key, value in defaults_true_continuum.items():
+            if key not in config["expected flux"]:
+                config["expected flux"][key] = str(value)
+        for key, value in defaults_desisim_data.items():
+            if key not in config["data"]:
+                config["data"][key] = str(value)
+        expected_flux = TrueContinuum(config["expected flux"])
+
+        var_lss = expected_flux.get_var_lss(10**Forest.log_lambda_grid)[:-1]
```
Comment (iprafols) [inline py/picca/tests/delta_extraction/expected_flux_tests.py:893]: I would get the `log_lambda_grid` from the test file. This way if we change the default values for this variable (in `setup_forest`) the test will still work

Comment (iprafols) [reply py/picca/tests/delta_extraction/expected_flux_tests.py:893]: Also this means we can drop the `Forest` import at the top

File: py/picca/tests/delta_extraction/expected_flux_tests.py [hunk 3]
```diff
@@ -805,6 +812,442 @@ def test_expected_flux(self):
...
+
+        for key, value in defaults_true_continuum.items():
+            if key not in config["expected flux"]:
+                config["expected flux"][key] = str(value)
+        for key, value in defaults_desisim_data.items():
+            if key not in config["data"]:
+                config["data"][key] = str(value)
+        expected_flux = TrueContinuum(config["expected flux"])
+
+        log_lambda_ = np.arange(3.5563025007672873, 3.7403626894942437, 0.0003)
```
Comment (iprafols) [inline py/picca/tests/delta_extraction/expected_flux_tests.py:929]: rename to `log_lambda`

File: py/picca/tests/delta_extraction/expected_flux_tests.py [hunk 4]
```diff
@@ -805,6 +812,442 @@ def test_expected_flux(self):
...
+        np.testing.assert_equal(var_lss, var_lss_target)
+        np.testing.assert_equal(mean_flux, mean_flux_target)        
+
+
+    def test_true_continuum_read_true_continuum(self):
+        """Test reading true continuum from mocks"""
+        # setup Forest variables; case: linear wavelength solution
+        setup_forest("lin", pixel_step=2.4)
+
+        data_dir = Path(__file__).parent / "data"
```
Comment (iprafols) [inline py/picca/tests/delta_extraction/expected_flux_tests.py:948]: same here

File: py/picca/tests/delta_extraction/expected_flux_tests.py [hunk 5]
```diff
@@ -805,6 +812,442 @@ def test_expected_flux(self):
...
+                    forest.continuum,
+                    np.loadtxt(data_dir / "true_true_cont_log.txt")
+                )
+
+    def test_true_continuum_expected_flux(self):
+        """Test method compute expected flux for class TrueContinuum"""
+        # setup Forest variables; case: logarithmic wavelength solution
+        setup_forest("log", rebin=3)
+
+        data_dir = Path(__file__).parent / "data"
```
Comment (iprafols) [inline py/picca/tests/delta_extraction/expected_flux_tests.py:1022]: same here

File: py/picca/tests/delta_extraction/expected_flux_tests.py [hunk 6]
```diff
@@ -805,6 +812,442 @@ def test_expected_flux(self):
...
+                config["data"][key] = str(value)
+        data = DesisimMocks(config["data"])
+        expected_flux = TrueContinuum(config["expected flux"])
+
+        # compute the expected flux
+        expected_flux.compute_expected_flux(data.forests)
+
+        # check the results
+        for iteration in range(1,5):
+            if "REPLACE_TEST_VALUES" in os.environ and os.environ["REPLACE_TEST_VALUES"] == "Y":
```
Comment (iprafols) [inline py/picca/tests/delta_extraction/expected_flux_tests.py:1053]: I would not add this copying thing here. The point of manually copying the files when we really need to update the test suite is to make sure we don't do it automatically by mistake (e.g. if the environment variable was set for some other reason)

File: py/picca/tests/delta_extraction/expected_flux_tests.py [hunk 7]
```diff
@@ -805,6 +812,442 @@ def test_expected_flux(self):
...
+            if key not in config["data"]:
+                config["data"][key] = str(value)
+        data = DesisimMocks(config["data"])
+        expected_flux = TrueContinuum(config["expected flux"])   
+
+        expected_flux.compute_expected_flux(data.forests)
+
+        # check the results
+        for iteration in range(1,5):   
+            if "REPLACE_TEST_VALUES" in os.environ and os.environ["REPLACE_TEST_VALUES"] == "Y":
```
Comment (iprafols) [inline py/picca/tests/delta_extraction/expected_flux_tests.py:1099]: same here

File: py/picca/tests/delta_extraction/expected_flux_tests.py [hunk 8]
```diff
@@ -805,6 +812,442 @@ def test_expected_flux(self):
...
+                
+            self.compare_fits(
+                str(test_file).replace(".fits", f"_iteration{iteration}.fits"),
+                str(out_file).replace(".fits", f"_iteration{iteration}.fits"))        
+        if "REPLACE_TEST_VALUES" in os.environ and os.environ["REPLACE_TEST_VALUES"] == "Y":
+            shutil.copyfile(out_file, test_file)
+        self.compare_fits(test_file, out_file)
+
+    def test_true_cont_compute_mean_cont_lin(self):
+        """Test method compute_mean_cont_lin for class TrueContinuum"""
```
Comment (iprafols) [inline py/picca/tests/delta_extraction/expected_flux_tests.py:1113]: the method is called `compute_mean_cont`

File: py/picca/tests/delta_extraction/expected_flux_tests.py [hunk 9]
```diff
@@ -805,6 +812,442 @@ def test_expected_flux(self):
...
+        if "REPLACE_TEST_VALUES" in os.environ and os.environ["REPLACE_TEST_VALUES"] == "Y":
+            shutil.copyfile(out_file, test_file)
+        self.compare_fits(test_file, out_file)
+
+    def test_true_cont_compute_mean_cont_lin(self):
+        """Test method compute_mean_cont_lin for class TrueContinuum"""           
+        # setup Forest variables; case: linear wavelength solution
+        setup_forest("lin", pixel_step=2.4)
+
+        data_dir = Path(__file__).parent / "data"
```
Comment (iprafols) [inline py/picca/tests/delta_extraction/expected_flux_tests.py:1117]: same here

File: py/picca/tests/delta_extraction/expected_flux_tests.py [hunk 10]
```diff
@@ -805,6 +812,442 @@ def test_expected_flux(self):
...
+                config["data"][key] = str(value)
+        data = DesisimMocks(config["data"])
+        expected_flux = TrueContinuum(config["expected flux"])   
+
+        expected_flux.compute_expected_flux(data.forests)
+    
+        mean_cont = expected_flux.get_mean_cont(Forest.log_lambda_rest_frame_grid)
+        mean_cont_weight = expected_flux.get_mean_cont(Forest.log_lambda_rest_frame_grid)
+
+        if "REPLACE_TEST_VALUES" in os.environ and os.environ["REPLACE_TEST_VALUES"] == "Y":
```
Comment (iprafols) [inline py/picca/tests/delta_extraction/expected_flux_tests.py:1145]: same here

File: py/picca/tests/delta_extraction/expected_flux_tests.py [hunk 11]
```diff
@@ -805,6 +812,442 @@ def test_expected_flux(self):
...
+            mean_cont,
+            np.loadtxt(data_dir / "true_mean_cont_lin.txt")
+        )
+        np.testing.assert_equal(
+            mean_cont_weight,
+            np.loadtxt(data_dir / "true_mean_cont_weight_lin.txt")
+        )
+
+    def test_true_cont_compute_mean_cont_log(self):
+        """Test method compute_mean_cont_log for class TrueContinuum"""
```
Comment (iprafols) [inline py/picca/tests/delta_extraction/expected_flux_tests.py:1159]: same here

File: py/picca/tests/delta_extraction/expected_flux_tests.py [hunk 12]
```diff
@@ -805,6 +812,442 @@ def test_expected_flux(self):
...
+            mean_cont_weight,
+            np.loadtxt(data_dir / "true_mean_cont_weight_lin.txt")
+        )
+
+    def test_true_cont_compute_mean_cont_log(self):
+        """Test method compute_mean_cont_log for class TrueContinuum"""           
+        # setup Forest variables; case: logarithmic wavelength solution
+        setup_forest("log", rebin=3)
+
+        data_dir = Path(__file__).parent / "data"
```
Comment (iprafols) [inline py/picca/tests/delta_extraction/expected_flux_tests.py:1163]: same here

File: py/picca/tests/delta_extraction/expected_flux_tests.py [hunk 13]
```diff
@@ -805,6 +812,442 @@ def test_expected_flux(self):
...
+        data = DesisimMocks(config["data"])
+        expected_flux = TrueContinuum(config["expected flux"])
+
+        # compute the expected flux
+        expected_flux.compute_expected_flux(data.forests)
+
+        mean_cont = expected_flux.get_mean_cont(Forest.log_lambda_rest_frame_grid)
+        mean_cont_weight = expected_flux.get_mean_cont(Forest.log_lambda_rest_frame_grid)
+
+        if "REPLACE_TEST_VALUES" in os.environ and os.environ["REPLACE_TEST_VALUES"] == "Y":
```
Comment (iprafols) [inline py/picca/tests/delta_extraction/expected_flux_tests.py:1195]: same here

**Review by iprafols — CHANGES_REQUESTED** [review]
There are some things that need changing, note that I added the "same here" comment where I had already added a related comment before not to repeat too much text. Please double check that I did not miss any of such repetitions

**Comment by cramirezpe** [comment]
@iprafols I think we can already merge this. Is there anything I need to do before?

**Review by iprafols — APPROVED** [review]

**Pull request merged**